### PR TITLE
Enable knowledge distillation for AutoMM

### PR DESCRIPTION
**Pull request by zhiqiangdon**
1. Design the distillation API for AutoMMPredictor. Knowledge distillation reuses the .fit() function. Here is an example to use distillation.
```
from autogluon.text.automm import AutoMMPredictor
teacher_predictor = AutoMMPredictor(label="label_column").fit(train_data)
student_predictor = AutoMMPredictor(label="label_column").fit(
    train_data, 
    hyperparameters=student_and_distiller_hparams, 
    teacher_predictor=teacher_predictor,
)
```
2. Add distillation loops for training and validation.
3. Add a setup function in AutoMMPredictor to prepare for distillation.
4. Reduce the memory usage for top k checkpoints averaging.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1670-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1670/2/index.html

**Comment by sxjscience** [comment]
@Cli212 @szha We added model distillation in AutoGluon. This implements part of the Distiller framework.

**Comment by jwmueller** [comment]
Nice! May be interesting to see if there exist datasets where student model whose inputs are only a subset of modalities can still retain good performance of teacher model trained on all modalities.

**Comment by sxjscience** [comment]
> Nice! May be interesting to see if there exist datasets where student model whose inputs are only a subset of modalities can still retain good performance of teacher model trained on all modalities.

Yes, we are doing the experiment on distilling to a student model that is operating on a less number of modalities than the teacher

**Comment by szha** [comment]
Job PR-1670-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1670/4/index.html

**Comment by szha** [comment]
Job PR-1670-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1670/5/index.html

File: text/src/autogluon/text/automm/predictor.py [hunk 1]
```diff
@@ -578,91 +756,127 @@ def _fit(
...
+                minmax_mode=minmax_mode,
+                is_distill=is_distill,
+                config=config,
+                val_df=val_df,
+                validation_metric_name=validation_metric_name,
+            )
+        else:
+            sys.exit(
+                f"Training finished, exit the process with global_rank={trainer.global_rank}..."
             )
```
Comment (gradientsky) [inline text/src/autogluon/text/automm/predictor.py:772]: Is this happening in a sub-process? Is there a clean way to exit the process without calling `sys.exit`?

File: text/src/autogluon/text/automm/predictor.py [hunk 2]
```diff
@@ -423,6 +443,128 @@ def fit(
...
+        logger.debug("setting up distillation...")
+        if isinstance(teacher_predictor, str):
+            teacher_predictor = AutoMMPredictor.load(teacher_predictor)
+
+        # verify that student and teacher configs are consistent.
+        assert self._problem_type == teacher_predictor._problem_type
+        assert self._label_column == teacher_predictor._label_column
+        assert self._eval_metric_name == teacher_predictor._eval_metric_name
+        assert self._output_shape == teacher_predictor._output_shape
+        assert self._validation_metric_name == teacher_predictor._validation_metric_name
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:485]: You can refer to here: https://github.com/awslabs/autogluon/blob/master/text/src/autogluon/text/automm/utils.py#L60

Basically, we use `torchmetrics` as the validation metric, which is required by pytorch lightning. But some `torchmetrics` may have bugs, e.g, r2.

File: text/src/autogluon/text/automm/predictor.py (hunk 1 above)
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:772]: For `ddp` and `ddp_spawn`, each GPU has its process. Refer to here: https://pytorch-lightning.readthedocs.io/en/1.4.3/advanced/multi_gpu.html#distributed-data-parallel

File: text/src/autogluon/text/automm/predictor.py [hunk 3]
```diff
@@ -423,6 +443,128 @@ def fit(
...
+                if n == teacher_predictor._model.prefix:
+                    teacher_predictor._model.prefix = new_name
+                else:
+                    assert isinstance(teacher_predictor._model.model, nn.ModuleList)
+                    for per_model in teacher_predictor._model.model:
+                        if n == per_model.prefix:
+                            per_model.prefix = new_name
+                            break
+                # modify data processor prefix
+                for _, per_modality_processors in teacher_predictor._data_processors.items():
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:505]: sure.

File: text/src/autogluon/text/automm/predictor.py [hunk 4]
```diff
@@ -423,6 +443,128 @@ def fit(
...
+        # verify that student and teacher configs are consistent.
+        assert self._problem_type == teacher_predictor._problem_type
+        assert self._label_column == teacher_predictor._label_column
+        assert self._eval_metric_name == teacher_predictor._eval_metric_name
+        assert self._output_shape == teacher_predictor._output_shape
+        assert self._validation_metric_name == teacher_predictor._validation_metric_name
+
+        # if teacher and student have duplicate model names, change teacher's model names
+        # we don't change student's model names to avoid changing the names back when saving the model.
+        teacher_model_names = []
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:488]: sure.

File: text/src/autogluon/text/automm/predictor.py [hunk 5]
```diff
@@ -578,91 +756,127 @@ def _fit(
...
+                ingredients = [top_k_model_paths[0]]
+                self._model = self._load_state_dict(
+                    model=model,
+                    path=top_k_model_paths[0],
+                    prefix=prefix,
+                )
+                best_score = self.evaluate(val_df, [validation_metric_name])[validation_metric_name]
+                for i in range(1, len(top_k_model_paths)):
+                    cand_avg_state_dict = average_checkpoints(
+                        checkpoint_paths=ingredients + [top_k_model_paths[i]],
```
Comment (gradientsky) [inline text/src/autogluon/text/automm/predictor.py:841]: I know it's not in the original paper, but should we try to evaluate all the pre-selected models here through multiple iterations (including already added) and add only the best one improving the soup? Considering models already selected in previous iterations would allow to dynamically adjust soup component weights (models added: A-B-C-A results in 0.5-0.25-0.25 weights for models A-B-C respectively).

**Comment by szha** [comment]
Job PR-1670-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1670/6/index.html

File: text/src/autogluon/text/automm/optimization/lit_distiller.py [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+            return self.validation_metric(logits.squeeze(dim=1), label)
+
+    def _shared_step(
+            self,
+            batch: dict,
+    ):
+        student_output = self.student_model(batch)
+        self.teacher_model.eval()
+        with torch.no_grad():
+            teacher_output = self.teacher_model(batch)
```
Comment (gradientsky) [inline text/src/autogluon/text/automm/optimization/lit_distiller.py:232]: Does it makes sense to cache teacher outputs? If we do this more than one epoch, this would save compute.

File: text/src/autogluon/text/automm/optimization/lit_distiller.py [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+                lr_decay=self.hparams.lr_decay,
+                weight_decay=self.hparams.weight_decay,
+            )
+        else:
+            logger.debug("applying single learning rate...")
+            grouped_parameters = apply_single_lr(
+                model=self.student_model,
+                lr=self.hparams.lr,
+                weight_decay=self.hparams.weight_decay,
+            )
```
Comment (gradientsky) [inline text/src/autogluon/text/automm/optimization/lit_distiller.py:325]: Same parameters passed into all three branches - consider moving those to dict - improves readability and reducing errors possibility:
```python
args = dict(
    model=self.student_model,
    lr=self.hparams.lr,
    weight_decay=self.hparams.weight_decay,
)
if self.hparams.lr_choice == "two_stages":
    logger.debug("applying 2-stage learning rate...")
    grouped_parameters = apply_two_stages_lr(
        lr_mult=self.hparams.lr_mult,
        return_params=True,
        **args,
    )
elif self.hparams.lr_choice == "layerwise_decay":
    logger.debug("applying layerwise learning rate decay...")
    grouped_parameters = apply_layerwise_lr_decay(
        lr_decay=self.hparams.lr_decay,
        **args,
    )
else:
    logger.debug("applying single learning rate...")
    grouped_parameters = apply_single_lr(
        **args,
    )
```

File: text/src/autogluon/text/automm/optimization/lit_distiller.py [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+        self.custom_metric_func = custom_metric_func
+
+    def _compute_hard_label_loss(
+            self,
+            output: dict,
+            label: torch.Tensor,
+    ):
+        loss = 0
+        for _, per_output in output.items():
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/lit_distiller.py:164]: Thanks.

File: text/src/autogluon/text/automm/predictor.py [hunk 9]
```diff
@@ -436,39 +578,75 @@ def _fit(
...
-            loss_func=loss_func,
-            validation_metric=validation_metric,
-            validation_metric_name=validation_metric_name,
-            custom_metric_func=custom_metric_func,
-            efficient_finetune=OmegaConf.select(config, 'optimization.efficient_finetune'),
+            val_data=val_df,
         )
+        is_distill = teacher_model is not None
+        if is_distill:
+            task = LitModuleForDistiller(
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:607]: sure.

File: text/src/autogluon/text/automm/predictor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:488]: done.

File: text/src/autogluon/text/automm/predictor.py [hunk 10]
```diff
@@ -436,39 +578,75 @@ def _fit(
...
+                lr=config.optimization.learning_rate,
+                lr_decay=config.optimization.lr_decay,
+                end_lr=config.optimization.end_lr,
+                lr_mult=config.optimization.lr_mult,
+                weight_decay=config.optimization.weight_decay,
+                warmup_steps=config.optimization.warmup_steps,
+                loss_func=loss_func,
+                validation_metric=validation_metric,
+                validation_metric_name=validation_metric_name,
+                custom_metric_func=custom_metric_func,
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:646]: Great suggestion!

File: text/src/autogluon/text/automm/optimization/lit_distiller.py (hunk 7 above)
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/lit_distiller.py:325]: Thanks! Done.

File: text/src/autogluon/text/automm/predictor.py (hunk 5 above)
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:841]: Make sense. We can compare the performance on our benchmark datasets. If doing in this way, the number of iterations may become a new hyper-parameter since it doesn't need to be `len(top_k_model_paths)`. If we use `while True` instead, we may fall into dead loops for the condition operators `operator.le` and `operator.ge`.

File: text/src/autogluon/text/automm/optimization/lit_distiller.py (hunk 6 above)
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/lit_distiller.py:232]: This is a trade-off between time and space. Caching teacher outputs can increase memory usage, which could be an issue for large datasets. In future, we will add distillation methods using hidden features. Caching the hidden features would have more memory cost.

**Comment by szha** [comment]
Job PR-1670-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1670/7/index.html

**Review by sxjscience — APPROVED** [review]
LGTM!

**Review by gradientsky — APPROVED** [review]
LGTM!

**Comment by szha** [comment]
Job PR-1670-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1670/8/index.html
